File: Python/0004_median_of_two_sorted_arrays/0004_median_of_two_sorted_arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        length = len(nums1) + len(nums2)
        
        if length % 2 == 0:
            return (self.findKth(nums1, nums2, 0, len(nums1), 0, len(nums2), length // 2) + self.findKth(nums1, nums2, 0, len(nums1), 0, len(nums2), length // 2 + 1)) / 2.0
        else:
            return self.findKth(nums1, nums2, 0, len(nums1), 0, len(nums2), length // 2 + 1)
    
    def findKth(self, nums1, nums2, start1, len1, start2, len2, k):
        if len1 > len2:
            return self.findKth(nums2, nums1, start2, len2, start1, len1, k)
        
        if len1 == 0:
            return nums2[start2 + k - 1]
        
        if k == 1:
            return min(nums1[start1], nums2[start2])
        
        p1 = min(k // 2, len1)
        p2 = k - p1
        if nums1[start1 + p1 - 1] > nums2[start2 + p2 - 1]:
            return self.findKth(nums1, nums2, start1, len1, start2 + p2, len2 - p2, k - p2)
        elif nums1[start1 + p1 - 1] < nums2[start2 + p2 - 1]:
            return self.findKth(nums1, nums2, start1 + p1, len1 - p1, start2, p2, k - p1)
        else:
            return nums1[start1 + p1 - 1]

File: Java/0004_median_of_two_sorted_arrays/0004_median_of_two_sorted_arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int len = nums1.length + nums2.length;
        
        if (len % 2 == 0) {
            return (findKth(nums1, nums2, 0, nums1.length, 0, nums2.length, len / 2) + findKth(nums1, nums2, 0, nums1.length, 0, nums2.length, len / 2 + 1)) / 2.0f;
        }
        else {
            return findKth(nums1, nums2, 0, nums1.length, 0, nums2.length, len / 2 + 1);
        }
    }
    
    public double findKth(int[] nums1, int[] nums2, int start1, int len1, int start2, int len2, int k) {
        if (len1 > len2) {
            return findKth(nums2, nums1, start2, len2, start1, len1, k);
        }
        
        if (len1 == 0) {
            return nums2[start2 + k - 1];
        }
        
        if (k == 1) {
            return Math.min(nums1[start1], nums2[start2]);
        }
        
        int p1 = Math.min(k / 2, len1);
        int p2 = k - p1;
        if (nums1[start1 + p1 - 1] > nums2[start2 + p2 - 1]) {
            return findKth(nums1, nums2, start1, len1, start2 + p2, len2 - p2, k - p2);
        }
        else if (nums1[start1 + p1 - 1] < nums2[start2 + p2 - 1]) {
            return findKth(nums1, nums2, start1 + p1, len1 - p1, start2, p2, k - p1);
        }
        else {
            return nums1[start1 + p1 - 1];
        }
    }
}